## Title Of The Paper
***

XXX

**Authors**:

**Journal**:

**Year**:

**Abstract**:


<div style="text-align:center"><img src="https://sismo.app/wp-content/uploads/2019/02/under-construction-gif-11.gif" width="250"/></div>

[Phylogenetic Tree](http://XXX)

### Saccharomyces strains

Request information for Saccharomyces strains used in this study can be found in Table SXXX.

### Raw Reads

The Illumina sequencing reads for this project has been deposed in the Short Reads Archive (SRA) under bioproject PRJNA475869

### Saccharomyces genomes

[Description of the sequencing and libraries]

New nuclear, mitochondrial and genome assemblies for different *Saccharomyces* lineages assembled by using Paired-End and Mate Pair Illumina libraries. FIles were deposited to the European Nucleotide Archive under bioproject PRJEB48264.

Strain |Species|Pop| ENA Accession no
-------|-------|---|------------------
yHAB335|*S.cerevisiae*|CHN IV|[GCA_918239655](LINK "GCA_918239655")
yHDPN23|*S.paradoxus*|Am-B|[GCA_918221835](LINK "GCA_918221835")
yHDPN24|*S.paradoxus*|Am-C|[GCA_918281175](LINK "GCA_918281175")
yHDPN32|*S.paradoxus*|EU/Am-A|[GCA_918265645](LINK "GCA_918265645")
IFO1815|*S.mikatae*|Asia-A|[GCA_918250775](LINK "GCA_918250775")
yHAB336|*S.mikatae*|Asia-B|[GCA_918272125](LINK "GCA_918272125")
IFO1802|*S.kudriavzevii*|Asia-A|[GCA_918252685](LINK "GCA_918252685")
IFO1803|*S.kudriavzevii*|Asia-B|[GCA_918257225](LINK "GCA_918257225")
yHKS300|*S.kudriavzevii*|EU|[GCA_918282095](LINK "GCA_918282095")
ZP591|*S.kudriavzevii*|EU|[GCA_918268175](LINK "GCA_918268175")
ZP960|*S.arboricola*|Oceania|[GCA_918268255](LINK "GCA_918268255")
CDFM21L.1|*S.eubayanus*|HOL|[GCA_918271895](LINK "GCA_918271895")
yHAB94|*S.eubayanus*|PA-A|[GCA_918265725](LINK "GCA_918265725")
yHCT69|*S.eubayanus*|PA-B|[GCA_918298305](LINK "GCA_918298305")
yHCT99|*S.eubayanus*|PA-A|[GCA_918246195](LINK "GCA_918246195")
yHRVM108|*S.eubayanus*|HOL|[GCA_918265445](LINK "GCA_918265445")
CBS7001|*S.uvarum*|HOL-EU|[GCA_918261815](LINK "GCA_918261815")
yHAB60|*S.uvarum*|HOL/SA-A|[GCA_918251765](LINK "GCA_918251765")
yHAB447|*S.uvarum*|HOL/SA-A|[GCA_918298155](LINK "GCA_918298155")
yHAB482|*S.uvarum*|HOL/SA-A|[GCA_918264245](LINK "GCA_918264245")
yHAB521|*S.uvarum*|SA-B|[GCA_918280685](LINK "GCA_918280685")
yHCT77|*S.uvarum*|HOL-NA|[GCA_918221285](LINK "GCA_918221285")
ZP964|*S.uvarum*|Australasia|[GCA_918269355](LINK "GCA_918269355")


Corrections or genome assemblies used in the recent work. Files were compressed using tar, to uncompress them use `tar -xvzf FILENAME`

Strain |Species|Pop| Nuclear | mtDNA | 2 micron plasmid
-------|-------|---|------------------|-------|-----------------
NCYC3947|*S.jurei*|EU|[Naseebetal2018](ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/900/290/405/GCA_900290405.1_SacJureiUoM1 "Naseebetal2018")|[Naseebetal2018(Corrected)](https://github.com/PerisD/Sac2.0/blob/master/MtGENOMES/NCYC3947.mtgenome.gz "Naseebetal2018(Corrected)")|Incomplete
CBS10644|*S.arboricola*|Asia-A|[Liti et al 2013](https://yjx1217.github.io/Yeast_PacBio_2016/data/ "YPRP")|[Suloetal2017](https://github.com/PerisD/Sac2.0/blob/master/MtGENOMES/CBS10644.mtgenome.gz "Suloetal2017")|In the nuclear genome


Other genome assemblies from recent papers stored in Github:
2. [Yue et al 2017](https://yjx1217.github.io/Yeast_PacBio_2016/data/ "YPRP"): CBS432, DBVPG6044, DBVPG6765, N44, SK1, UFRJ50816, UWOPS03-461.4, UWOPS91-917.1, Y12, YPS128, YPS138

The rest of genome assemblies for *Saccharomyces* strains, using just Paired-End Illumina reads, are compressed in this [file](https://github.com/PerisD/Sac2.0/blob/master/NGENOMES/other.assemblies.gz "OtherAssemblies")

*COX2* and *COX3* sequences were deposited in GenBank under accession nos. MH813536-MH813939.

*GAL* genes that were Sanger sequenced were deposited in GenBank under accession nos. OL660614-OL660618.

### Source data

Raw data, alignments and results were deposited in [Dryad](doi:10.5061/dryad.rxwdbrv9p "Dryad repository"). The repository contains:

1. Alignments of 2um plasmid genes *REP1* and *REP2* used in Figure XXX.

2. ASTRAL run and outputs to generate Figure XXX.

3. MrBayes and BUCKy raw data and results to generate Figure XXX.

4. Single copy ortholog gene and protein alignments annotated using BUSCO for animals and *Saccharomyces* and used to generate the Figure XXX.

5. GAL/MEL pathway genes and their encoded protein alignments, together with the IQTree log and Maximum Likelihood tree files. This dataset was used to generate Figure XXX.

6. Alignemtns of annotated genes and their proteins with Yeast Genome Annotation Pipeline (YGAP) and IQtree log and Maximum Likelihood tree files. This dataset was used to generate Figure XXX.

7. Individual mitochondrial gene alignments used to generate Figure XXX. COX2 and COX3 Sanger sequenced alignments to generate Figure XXX.

8. iWGS genome assemblies for those strains with just Paired-end Illumina reads.

9. Variant Call Format files necessary for population genomic analyses resulting in Figures XXX.

10. Optican density 600nm values, 96-well plate pictures and additional information for each media condition run necessary to generate Figures XXX.

### Used command lines
For in-house scripts, please check instructions by calling reading the script with a text editor (my first scripts) or by running ```python script.py --help``` for the new scripts.
If it is not specified, python scripts were run with python2 version 6+

#### Illumina reads download and trimming
```
[INPUTFile]: input file
[INPUTF]: path to the input folder containing illumina reads
[OUTPUTF]: output folder to store results
[READ]: read name
[FQEXTENSION]: extension name of the illumina read

python download_SRA_serially.py [INPUTFile] [OUTPUTF] YES
python SRA_rename_v1.4.py [INPUTF] [FQEXTENSION] [INPUTF] [OUTPUTF]

Paired-ended illumina reads
fastqc -o [OUTPUTF] -f fastq [READ]_1.fq
java -jar trimmomatic-0.33.jar PE -threads 4 -phred33 -trimlog analytics.txt [READ]_1.fq [READ]_2.fq [READ]_1.tm.fq [READ]_1.tm-se.fq [READ]_2.tm.fq [READ]_2.tm-se.fq ILLUMINACLIP:TruSeq2-PE.fa:2:30:10 TRAILING:3 MINLEN:25
fastqc -o [OUTPUTF] -f fastq [READ]_1.tm.fq

Mate pair illumina reads
nextclip --input_one [READ]_1.fq --input_two [READ]_2.fq --output_prefix trimmed --log log.txt --trim_ends 5
```

#### Quick assesment of phylogenetic position with AAF (Alignment and Assembly Free) 
```
[INPUTF]: path to the input folder containing a strain folder with the PE illumina reads for that strain
[KMER]: Number of kmers to be used
[RUNNAME]: output file name

python2.7 AAF_pipeline-kmer.py -i [INPUTF] -o [RUNNAME] -k [KMER]
Convert the generated infile file to MEGA format and open it in MEGA to generate a Neighbor-Joining phylogenetic tree.
```

#### Assemblies and quality check with iWGS
```
[NAME].ctl: describe the information about PE and MP illumina reads, assemblers to use, checkpoints to perform and quality assessment of the genome assemblies.
./iWGS -s [NAME].ctl --Real
```

#### Ultrascaffolding PE+MP genome assemblies
```
[GENOMEANAME]: genome assembly name
[INPUTF]: input folder
[INPUTFile]: input file

python runMummer_serially.py [INPUTFile]
Generated coords file is useful in the next step

Rscript --vanilla reorderContigsUsingNucmer.R [GENOMEANAME].coords [GENOMEANAME].scaffolds.fa
A file called remap.coords.result.csv is generated with the information about scaffolds mapping to chromosomes and if they were reverse complemented. This is useful information for manual curation in Geneious.

python combining_orderedScaffolds.py [INPUTF]
python generating_FinalAssembly.py -l 10000 -i [INPUTF]
Some strains will need manual curation in Geneious. After Geneious curation, run mummer for confirmation.
python runMummer_serially.py [INPUTFile]
```

#### mtDNA asembly pipeline
```

```

####  Large structural variation (nuclear genome and mitochondrial genome)
```

```

#### 
```

```

#### 
```

```

#### 
```

```

#### 
```

```

#### 
```

```

#### 
```

```

***

### Contact

[<img src="https://www.uv.es/perisnav/Index/twitter-logo.png" width="50"/>](https://linktr.ee/PerisD)

[<img src="https://www.uv.es/perisnav/images/Mitogression.png" width="60"/>](https://www.uv.es/perisnav/)

