__author__ = 'Peris'

#!/usr/bin/env python
from Bio import SeqIO
import sys
import argparse
import os
import glob
import shutil

helptext="""
This script will check all the alignments generated by splitFASATAintoWindows.py can be opened by PopGenome
Authors: Quinn Langdon modified by Peris UW-Madison, Dept Genetics & IATA-CSIC
"""

parser = argparse.ArgumentParser(description=helptext,formatter_class=argparse.RawTextHelpFormatter)
parser.add_argument("-i","--inputFolder", help="output folder, default wholeGenomeXRegions/", type = str, default = 'wholeGenomeXRegions/')

args = parser.parse_args()

temp_folder = "temp_folder"

if not os.path.exists(temp_folder):
	os.makedirs(temp_folder)

list_files = glob.glob(args.inputFolder+"*.fasta")

for i in list_files:
	os.makedirs(temp_folder+'/'+i)
	shutil.copy(i,temp_folder+'/'+i)

trimmed_alignments = open('trimmed_alignments.txt','w')
for i in list_files:
	outFile = temp_folder+"/"+"tmp_log_file_"+i.split('/')[-1]+".txt"
	popGenome_cm1 = "Rscript ~/software/scripts/checkPopgenome_Rmodule.R " + temp_folder+"/"+i + " &> "+ outFile
	os.system(popGenome_cm1)
	if "Error" in open(outFile, 'r').read():
		trimmed_alignments.write(temp_folder+"/"+i)
trimmed_alignments.close()

print "DONE!"